Added two option to Quaternion.to_euler

evbernardes opened the pull request:
#### References to other Issues or PRs
Little follow up to PR #24358


#### Brief description of what is fixed or changed
Added two options to `to_euler`.

The first uses $\operatorname{atan_2} (b,a) \pm \operatorname{atan_2} (d,c) = \operatorname{atan_2} (bc\pm ad, ac\mp bd)$ to assemble the first and third angles into single `atan2` expressions instead of a sum or difference of two `atan2` expressions. Can be useful to compare with existing conversion formulas on the internet.

The second one uses an equivalent formula for the second angle that uses `acos2`. The only advantage this has it to avoid the use of square roots.
#### Other comments


#### Release Notes

<!-- BEGIN RELEASE NOTES -->
* algebras
  * Added two options to Quaternion.to_euler.
<!-- END RELEASE NOTES -->

evbernardes committed:
added two options to to_euler

evbernardes committed:
fixed code quality

evbernardes committed:
fixed mailmap

oscarbenjamin commented on .mailmap:
```diff
@@ -535,6 +535,7 @@ Ethan Ward <etkewa@gmail.com> eward <eward@sunbelt-medical.com>
 Eva Charlotte Mayer <eva-charlotte.mayer@posteo.de>
 Eva Tiwari <eva.tiwari@gmail.com> Eva <eva.tiwari@gmail.com>
 Evandro <15084103+evbernardes@users.noreply.github.com>
+Evandro Bernardes <evbernardes@gmail.com> Evandro Bernardes <s.evandro@hotmail.com>
 Evandro Bernardes <evbernardes@gmail.com> evbernardes <s.evandro@hotmail.com>
```
Is `Evandro <15084103+evbernardes@users.noreply.github.com>` the same person as the other entries?

If so then the first of these lines should be changed to:
```
Evandro Bernardes <evbernardes@gmail.com> Evandro <15084103+evbernardes@users.noreply.github.com>
```
Currently this looks like it's referring to two different authors hence the authors test shows:
```
The following authors will be added to the AUTHORS file at the time of
the next SymPy release.
...
	Evandro <15084103+evbernardes@users.noreply.github.com>
	Evandro Bernardes <evbernardes@gmail.com>
```

evbernardes replied:
Oh my, I keep getting this wrong. Thanks again!

evbernardes committed:
fixed mailmap again

smichr commented:
A test should be added for this.

evbernardes commented:
> A test should be added for this.

Can I add some numerical tests instead of a symbolic one? I wasn't sure how to use sympy's simplifications to show the general relationship.

smichr commented:
> how to use sympy's simplifications

You should be able to use `a.equals(b)` -- not always, but it will often work.

evbernardes commented:
> > how to use sympy's simplifications
> 
> You should be able to use `a.equals(b)` -- not always, but it will often work.

I tried that, didn't work sadly.

evbernardes committed:
added numerical tests

evbernardes commented:
Alright, this is ugly, but I implemented some numerical tests.

evbernardes committed:
minor edit

smichr commented:
```python
>>> from sympy import this
...
Coverage by testing matters.
Smart tests are better than random tests.
But random tests sometimes find what your smartest test missed.
...
```
My preference for these sorts of tests is to 

1. do random numerical comparison to flush out any errors
2. once symbolics are believed to be correct, test a single symbolic form with unique args that don't serendipitously cancel

That way we don't have to keep retesting randomly what we believe to be true and have tested to be true. If someone changes the form that is output, they would have to repeat this process to prove that the new arrangement is correct.

In this case, perhaps a test better than random test is to cover all a,b,c,d with 
```python
for a,b,c,d in cartes(*[[-1,0,1]]*4):
```
since that would exercise all options for atan2 in terms of quadrants. But then once that stress test is done, only the symbolic solution could be affirmed. You could leave a "stress test" commit and then "symbolic test" commit afterwards so one could see what was done to test the symbolics.

evbernardes committed:
added zero quaternion check

evbernardes committed:
added checks for numerical singularity

evbernardes committed:
added singularity checks

evbernardes commented:
Alright, implemented some new things:

1. Now, if the elements are numerical, I do singularity tests.
2. Implemented tests where I force some singularity tests, expect a warning and check if the result is what would be expected.
3. Removed random tests and Implemented a new test checking, for every possible sequence, using:
```
for elements in itertools.product([-1, 0, 1], repeat=4)
    q = Quaternion(*elements)`
```

evbernardes committed:
minor edit

evbernardes committed:
minor edit

evbernardes committed:
small code cleanup

evbernardes committed:
using internal sympy pytest

evbernardes committed:
added simplification

evbernardes committed:
changed is_Number to is_number

evbernardes commented on sympy/algebras/quaternion.py:
```diff
@@ -244,26 +271,49 @@ def to_euler(self, seq):
+        angle_test = angles[1].evalf()
+        case = 0
+        pi_f = pi.evalf()
+        if angle_test.is_Number:
```
Changed `is_Number` to `is_number` for now.

I don't think there is any numerical problems though, because either:
1.  `atan2` is called with two positive arguments, which only gives a value between `0` and `pi/2` and then is multiplied by 2.
2. `acos` is called with properly normalized values, which can only give results between `0` and `pi`.

In both cases, the result is assured to be real and between `0` and `pi`. 
This is how it is implemented in the numerical conversion in SciPy for example.

evbernardes replied:
If you still prefer the use of `case = Piecewise((1, Eq(angles[1], 0)), (2, Eq(angles[1], pi), (0, True))` though, can you guide me true that? I don't really know what you meant here!

smichr replied:
How obfuscated will be expressions for which angle[1] is 0 or pi? Perhaps you can just test `is_eq(angle[1], S.Zero):` and `is_eq(angle[1], S.Pi)`?

evbernardes replied:
By `is_eq` do you mean `equals`?

smichr replied:
Since something like `if Eq(x, 10)` will not always evaluate you can't use it (except ackwardly) as a condition, so helpers like `is_eq` were added that can be used as `if is_eq(x, 10)`. `from sympy.core.relational import is_eq`.

evbernardes replied:
Alright I see it now! Updated accordingly.

evbernardes committed:
now testing with equals

evbernardes committed:
simplified test_to_euler

evbernardes commented:
Added a little simplification to `test.to_euler`. Instead of testing every one of the possible 24 sequences, I only test 8 now:

1. two asymmetrical: `zyx` (odd parity) and `zxy` (even parity)
2. two symmetrical: `zyz` (odd parity) and `zxz` (even parity)
3. the intrinsical equivalents of these four: `ZYX`, `ZXY`, `ZYZ` and `ZXZ`

This way the test is simpler, tests every possible case and is a lot faster.

evbernardes committed:
using is_eq

oscarbenjamin commented on sympy/algebras/quaternion.py:
```diff
@@ -244,26 +271,47 @@ def to_euler(self, seq):
+                angles[2] = atan2(b*c - a*d, a*c + b*d)
+
+        else:  # any degenerate case
+            warnings.warn('Singularity case, setting third angle to zero')
```
Please don't add warnings like this. It is better just to give an error or something.

oscarbenjamin commented:
Now the test suite gives:
```console
$ bin/test sympy/algebras/
======================================== test process starts ========================================
executable:         /home/oscar/current/sympy/sympy.git/38venv/bin/python  (3.8.9-final-0) [CPython]
architecture:       64-bit
cache:              yes
ground types:       gmpy 2.1.2
numpy:              1.23.5
random seed:        88073870
hash randomization: on (PYTHONHASHSEED=3239893863)

sympy/algebras/tests/test_quaternion.py[18] ................./home/oscar/current/sympy/sympy.git/sympy/algebras/quaternion.py:557: UserWarning: Singularity case, setting third angle to zero
  warnings.warn('Singularity case, setting third angle to zero')
```
There should be no warnings output when running the test suite. There are ways to suppress warnings when running the test suite but really it is just better if we don't use the warning system at all.

Why can't this case just be an error?

evbernardes commented:
Hey there @oscarbenjamin, this is because this isn't really an error.

In singularity/degenerate cases, the system is underdetermined, and there are infinite solutions (look up Euler angles gymbal lock).

What is usually done is that in this case we just set one of the angles to 0.

If the `warning` is a problem, it can be removed, but it shouldn't be an error.

evbernardes commented:
This comes from another PR I made first though: #24358

oscarbenjamin commented:
> If the `warning` is a problem, it can be removed, but it shouldn't be an error.

Okay, that seems reasonable.

I think it's generally better just to document how degenerate cases are handled or raise an error rather than give a warning. Warnings are very useful for showing that something is deprecated without immediately breaking it but I wouldn't use them in a situation where the user should not be expected to change their code.

evbernardes commented:
Ok, I see! 

I removed the warnings in a new PR, and if you want I can add a little sentence to the documentation tomorrow.
